src/EditorFeatures/Core/InheritanceMargin/AbstractInheritanceMarginService.cs

```diff
@@ -0,0 +1,177 @@
+        /// </summary>
+        protected abstract int GetIdentifierLineNumber(SourceText sourceText, SyntaxNode declarationNode);
+
+        public async ValueTask<ImmutableArray<InheritanceMemberItem>> GetInheritanceInfoAsync(
+            Document document,
+            TextSpan spanToSearch,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var allDeclarationNodes = GetMembers(root, spanToSearch);
+            if (allDeclarationNodes.IsEmpty)
+            {
+                return ImmutableArray<InheritanceMemberItem>.Empty;
+            }
+
+            var sourceText = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            var mappingService = document.Project.Solution.Workspace.Services.GetRequiredService<ISymbolMappingService>();
+            using var _ = ArrayBuilder<InheritanceMemberItem>.GetInstance(out var builder);
+
+            // Iterate all the members symbol to find each their inheritance chain information
+            foreach (var memberDeclarationNode in allDeclarationNodes)
+            {
+                var member = semanticModel.GetDeclaredSymbol(memberDeclarationNode, cancellationToken);
+                if (member != null && !member.IsStatic)
+                {
+                    var mappingResult = await mappingService.MapSymbolAsync(document, member, cancellationToken).ConfigureAwait(false);
+                    if (mappingResult != null)
+                    {
+                        if (mappingResult.Symbol is INamedTypeSymbol namedTypeSymbol)
+                        {
+                            // Find its baseTypes and subTypes.
+                            await AddInheritanceMemberItemsForNamedTypeAsync(
+                                mappingResult.Project.Solution,
+                                sourceText,
+                                memberDeclarationNode,
+                                namedTypeSymbol, builder, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        if (mappingResult.Symbol.IsOrdinaryMethod() || mappingResult.Symbol is IEventSymbol or IPropertySymbol)
+                        {
+                            // Find the implementing/implemented/overridden/overriding members
+                            await AddInheritanceMemberItemsForTypeMembersAsync(
+                                mappingResult.Project.Solution,
+                                sourceText,
+                                memberDeclarationNode,
+                                mappingResult.Symbol,
+                                builder,
+                                cancellationToken).ConfigureAwait(false);
```

Comment:
the declaration node and source text is passed in in both cases to determine the line number.  consider just pulling out that logic to this method and just passing in only the line number as that is all that is needed.

consider making this `builder.AddIfNotNull(await TryCreate...Item(...))`  and just make the helpers not take a builder but isntead return an item if they succeed.